command/agent/operator_endpoint.go

```diff
@@ -372,3 +375,91 @@ func (s *HTTPServer) snapshotSaveRequest(resp http.ResponseWriter, req *http.Req
 
 	return nil, codedErr
 }
+
+func (s *HTTPServer) snapshotRestoreRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	args := &structs.SnapshotRestoreRequest{}
+	s.parseWriteRequest(req, &args.WriteRequest)
+
+	var handler structs.StreamingRpcHandler
+	var handlerErr error
+
+	if server := s.agent.Server(); server != nil {
+		handler, handlerErr = server.StreamingRpcHandler("Operator.SnapshotRestore")
+	} else if client := s.agent.Client(); client != nil {
+		handler, handlerErr = client.RemoteStreamingRpcHandler("Operator.SnapshotRestore")
+	} else {
+		handlerErr = fmt.Errorf("misconfigured connection")
+	}
+
+	if handlerErr != nil {
+		return nil, CodedError(500, handlerErr.Error())
+	}
+
+	httpPipe, handlerPipe := net.Pipe()
+	decoder := codec.NewDecoder(httpPipe, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(httpPipe, structs.MsgpackHandle)
+
+	// Create a goroutine that closes the pipe if the connection closes.
+	ctx, cancel := context.WithCancel(req.Context())
+	defer cancel()
+	go func() {
+		<-ctx.Done()
+		httpPipe.Close()
+	}()
+
+	errCh := make(chan HTTPCodedError, 2)
+	go func() {
+		defer cancel()
+
+		// Send the request
+		if err := encoder.Encode(args); err != nil {
+			errCh <- CodedError(500, err.Error())
+			return
+		}
+
+		go func() {
+			var wrapper cstructs.StreamErrWrapper
+			bytes := make([]byte, 1024)
+
+			for {
+				n, err := req.Body.Read(bytes)
+				if n > 0 {
+					wrapper.Payload = bytes[:n]
+					err := encoder.Encode(wrapper)
+					if err != nil {
+						errCh <- CodedError(500, err.Error())
+						return
+					}
+				}
+				if err != nil {
+					wrapper.Payload = nil
+					wrapper.Error = &cstructs.RpcError{Message: err.Error()}
+					err := encoder.Encode(wrapper)
+					if err != nil {
+						errCh <- CodedError(500, err.Error())
+					}
+					return
+				}
+			}
+		}()
+
+		var res structs.SnapshotRestoreResponse
+		if err := decoder.Decode(&res); err != nil {
```

Comment:
Just for my understanding: this will block until we get the response? The sending/encoding goroutine will get EOF and exit if this fails and we close the `httpPipe`?

Reply:
Correct - either we get a valid response a valid `SnapshotRestoreResponse` here, or we get a parsing/EOF/pipe-closed.  In either case, we'll exit from this goroutine and cancel context.

This pattern of goroutine handling and error propagation is following previous examples of [FS handling](https://github.com/hashicorp/nomad/blob/v0.11.3/command/agent/fs_endpoint.go#L376-L420) and [Alloc exec](https://github.com/hashicorp/nomad/blob/master/command/agent/alloc_endpoint.go#L477-L533).